### ISSUE TITLE
Page reloads

### ISSUE THREAD

**Issue by peebles** [mjrussell/redux-auth-wrapper#108]
The user state is fetched from the server, along with a session cookie, on login using a async thunk.  Your examples address this nicely and everything works, until the user does a page reload.  The state gets nuked and the user is directed back to the login page because UserIsAuthenticated fails (no user state data).  However the session cookie is still valid.  I want to be able to call an async "/me" endpoint on the server which will return the logged in user if the cookie is still valid.  This should happen somehow during, or just before the UserIsAuthenticated check.  I do not understand the paragraph in the doc that says:

"After applying the configObject, UserAuthWrapper returns a function which can applied to a Component to wrap in authentication and authorization checks. The function also has the following extra properties:

onEnter(store, nextState, replace) (Function): Function to be optionally used in the onEnter property of a route."

Does this onEnter() also have a 4th arg, the "next()" that a normal onEnter function has that makes it async?  How do I use this "extra property"?  **"The function also has the following extra properties"**: I didn't know that functions **had** properties!

**Comment by mjrussell** [comment]
> I want to be able to call an async "/me" endpoint on the server which will return the logged in user if the cookie is still valid. This should happen somehow during, or just before the UserIsAuthenticated check

You should use the `authenticatingSelector` and `LoadingComponent` together to achieve this. You can set a flag in your reducer called `isLoading` or something like that which is what the `authenticatingSelector` points to. When you app mounts (can use `componentWillMount` of a top level component or some other method) you can dispatch an action that fetches `/me`. If this call is successful it will set the necessary data for the `authSelector and set `isLoading` to false.

The `LoadingComponent` will provide nice feedback for your users so you can show a spinner or something while this call is happening.

The `onEnter` is more useful in server side rendering than anything else in my opinion. The `onEnter` is used to add authentication checks to a React Router onEnter and doesn't have any async properties. So if you wanted to go that route you'd create your own onEnter function which uses `next`, makes the async fetch, and then calls the `onEnter` of the HOC.

> I didn't know that functions had properties

Technically in javascript, functions are just objects so you can tack on an additional property just like you would on an object.

**Comment by peebles** [comment]
I struggled a little bit, but I think its working now.  I ended up using your componentWillMount() suggestion.

I know there are a million ways to architect a react app (boo!) but this session cookie-based auth method between client and server has got to be pretty common, and handling page reload gracefully and properly would be a great additional example.  It could be faked by a setTimeout in componentWillMount() in the app.js of a new example.

I do like your project here.  I've been doing react now for more than a year, but have stuggled every project trying to implement a decent authentication workflow.  I have not found any tutorials suggesting how to do it.  Your project helps quite a bit.

If I knew more about redux middleware, that might be another way of trying to solve this.  

Thanks.

**Comment by peebles** [comment]
I have:
```javascript
<Route path="/expired" component={UserIsAuthenticated(PasswordExpiredPage)} />
```
On that page, the user will do something that will cause the user start to be scrubbed, forcing the new route to be "/login?redirect=%2Fexpired".  But in this case, I do not want the "redirect" query parameter to be added.  There may be other cases too where I would want to prevent the "redirect" param.

I tried adding a allowRedirectBack: function(...) to my UserIsAuthenticated, but that never gets called.  What am I missing?  Basically, depending on where the user was when the user became unauthenticated, I want to conditionally add (or not) the redirect param.

**Comment by peebles** [comment]
```javascript
const UserIsAuthenticated = UserAuthWrapper({
  authSelector: state => state.user, // how to get the user state
  authenticatingSelector: (state) => { return ( state.user && state.user.isLoading === true ? true : false ); },
  predicate: (user) => { return user.id ? true : false; },
  redirectAction: routerActions.replace, // the redux action to dispatch for redirect
  allowRedirectBack: (location, redirectPath) => {
    console.log( 'allow redirect:', location, redirectPath );
    return true;
  },
  wrapperDisplayName: 'UserIsAuthenticated' // a nice name for this auth check
});
```

**Comment by mjrussell** [comment]
Ah there isn't a release with the `allowRedirectBack` as a function yet. I'll try to get one out today

**Comment by peebles** [comment]
Status?

**Comment by mjrussell** [comment]
Sorry about that, I've been pretty swamped lately. Just released `v0.10.0` which includes the `allowRedirectBack` as a function as well as a bool.
